todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.java

```diff
@@ -122,65 +189,126 @@ public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
         super.onCreateOptionsMenu(menu, inflater);
     }
 
-    @Override
-    public void setLoadingIndicator(boolean active) {
-        if (active) {
-            mDetailTitle.setText("");
-            mDetailDescription.setText(getString(R.string.loading));
+    private void showTask(@NonNull Task task) {
+        String title = task.getTitle();
+        String description = task.getDescription();
+
+        if (title != null && title.isEmpty()) {
+            hideTitle();
+        } else {
+            showTitle(title);
+        }
+
+        if (description != null && description.isEmpty()) {
+            hideDescription();
+        } else {
+            showDescription(description);
         }
+        showCompletionStatus(task.isCompleted());
     }
 
-    @Override
-    public void hideDescription() {
-        mDetailDescription.setVisibility(View.GONE);
+    private void setLoadingIndicator(boolean active) {
+        Preconditions.checkNotNull(mLoadingProgress);
+
+        int visibility = active ? View.VISIBLE : View.GONE;
+        mLoadingProgress.setVisibility(visibility);
     }
 
-    @Override
-    public void hideTitle() {
-        mDetailTitle.setVisibility(View.GONE);
+    private void hideDescription() {
+        Preconditions.checkNotNull(mDetailDescription);
+        getDetailDescription().setVisibility(View.GONE);
     }
 
-    @Override
-    public void showDescription(String description) {
-        mDetailDescription.setVisibility(View.VISIBLE);
-        mDetailDescription.setText(description);
+    private void hideTitle() {
+        getDetailTitle().setVisibility(View.GONE);
     }
 
-    @Override
-    public void showCompletionStatus(final boolean complete) {
+    private void showDescription(String description) {
+        getDetailDescription().setVisibility(View.VISIBLE);
+        getDetailDescription().setText(description);
+    }
+
+    private void showCompletionStatus(final boolean complete) {
+        Preconditions.checkNotNull(mDetailCompleteStatus);
+
         mDetailCompleteStatus.setChecked(complete);
         mDetailCompleteStatus.setOnCheckedChangeListener(
                 new CompoundButton.OnCheckedChangeListener() {
                     @Override
                     public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                         if (isChecked) {
-                            mPresenter.completeTask();
+                            completeTask();
```

Comment:
You can do something similar what I mentioned in other PR, change this to observable